### PR TITLE
Ensure explicit output `dtype` for `pad_across_processes`

### LINKED ISSUE #3218: Incorrect type in output of `utils.pad_across_processes` when input is `torch.bool`
### System Info

```Shell
- `Accelerate` version: 1.1.0
- Platform: Linux-6.8.0-45-generic-x86_64-with-glibc2.35
- `accelerate` bash location: .venv/bin/accelerate
- Python version: 3.11.10
- Numpy version: 2.1.1
- PyTorch version (GPU?): 2.4.1+cu121 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- PyTorch MLU available: False
- PyTorch MUSA available: False
- System RAM: 755.50 GB
- GPU type: NVIDIA RTX 6000 Ada Generation
- `Accelerate` default config:
        Not found
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

Running the following code using `accelerate launch example.py`

```python3
import torch
from accelerate import Accelerator
from accelerate.utils import pad_across_processes

accelerator = Accelerator()

process_tensor = (torch.randn(2, 100*(accelerator.process_index + 1)) > 0).to(accelerator.device)
print(f"{process_tensor.shape = }, {process_tensor.dtype = }, {accelerator.process_index = }")

padded_tensor = pad_across_processes(process_tensor, dim=1)
print(f"{padded_tensor.shape = }, {padded_tensor.dtype = }, {accelerator.process_index = }")
```

On a machine with at least two GPUs will output (example for three GPUs):
```
process_tensor.shape = torch.Size([2, 300]), process_tensor.dtype = torch.bool, accelerator.process_index = 2
process_tensor.shape = torch.Size([2, 100]), process_tensor.dtype = torch.bool, accelerator.process_index = 0
process_tensor.shape = torch.Size([2, 200]), process_tensor.dtype = torch.bool, accelerator.process_index = 1

padded_tensor.shape = torch.Size([2, 300]), padded_tensor.dtype = torch.bool, accelerator.process_index = 2
padded_tensor.shape = torch.Size([2, 300]), padded_tensor.dtype = torch.int64, accelerator.process_index = 0
padded_tensor.shape = torch.Size([2, 300]), padded_tensor.dtype = torch.int64, accelerator.process_index = 1
```

The padded tensors have the incorrect data type of `torch.int64` and there is cross-device mismatch, which will further make downstream (e.g., `gather`) ops freeze and hard to debug.

### Expected behavior

The output tensor should have the same `dtype` on all devices, and it should be the same as the input `dtype`

### PR DESCRIPTION
**Pull request by mariusarvinte**
# What does this PR do?

Fixes #3218.

Current implementation casts `torch.bool` to `torch.int64` because of `+ pad_index`, where `pad_index` is `0` by default: https://github.com/huggingface/accelerate/blob/d0e80e51b68a2d39741ee948a9800d8298c8c95d/src/accelerate/utils/operations.py#L672

Adds a test case for checking that `torch.bool` is output with the same type.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [x] Did you write any new necessary tests?

## Who can review?

@muellerzr @BenjaminBossan @SunMarc

### CODE REVIEW

File: src/accelerate/utils/operations.py [hunk 1]
```diff
@@ -669,7 +669,7 @@ def _pad_across_processes(tensor, dim=0, pad_index=0, pad_first=False):
         old_size = tensor.shape
         new_size = list(old_size)
         new_size[dim] = max_size
-        new_tensor = tensor.new_zeros(tuple(new_size)) + pad_index
+        new_tensor = (tensor.new_zeros(tuple(new_size)) + pad_index).to(tensor.dtype)
```
Comment (BenjaminBossan) [inline src/accelerate/utils/operations.py:672]: Hmm, I'm wondering how safe this is in case that the tensor dtype cannot represent the new data. E.g. when `pad_index` is not 0 or 1, casting this to `bool` will result in a loss of information.

Comment (SunMarc) [reply src/accelerate/utils/operations.py:672]: +1, `pad_index` in the context of LLMs is usually `-100`.

Comment (mariusarvinte) [reply src/accelerate/utils/operations.py:672]: I don't think there's any loss of useful information per se, given that the data is assigned directly downstream 
https://github.com/huggingface/accelerate/blob/c0552c9012a9bae7f125e1df89cf9ee0b0d250fd/src/accelerate/utils/operations.py#L679

What this does change though is the actual pad value. Anything non-zero (e.g., -100) will result in padding with `True` for `bool`.

Does it make sense to always pad with `False` for `bool`?
